### PR TITLE
[Metrics] add bounded and unbounded mpsc channel wrappers

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

Majority of the implementation is adapted from the existing metered bounded channel. The main advantage of the new implementation is that it eliminates the need to create separate metrics for each channel.

Channels in `consensus-core` are migrated to the new monitored channel wrappers. This adds monitoring for unbounded channels.

## Test plan 

CI

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: crates/mysten-metrics/src/monitored_mpsc.rs [hunk 1]
```diff
@@ -0,0 +1,855 @@
...
+    pub async fn closed(&self) {
+        self.inner.closed().await
+    }
+
+    /// Attempts to immediately send a message on this `Sender`
+    /// Increments the gauge in case of a successful `try_send`.
+    pub fn try_send(&self, message: T) -> Result<(), TrySendError<T>> {
+        self.inner
+            .try_send(message)
+            // remove this unsightly hack once https://github.com/rust-lang/rust/issues/91345 is resolved
```
Comment (arun-koshy) [inline crates/mysten-metrics/src/monitored_mpsc.rs:53]: looks like the issue is resolved

**Review by arun-koshy — APPROVED** [review]
Nice!

File: crates/mysten-metrics/src/metered_channel.rs [hunk 2]
```diff
@@ -167,7 +167,9 @@ impl<'a, T> Permit<'a, T> {
 impl<'a, T> Drop for Permit<'a, T> {
     fn drop(&mut self) {
         // in the case the permit is dropped without sending, we still want to decrease the occupancy of the channel
-        self.gauge_ref.dec()
+        if self.permit.is_some() {
```
[the remark this answers is not in the document]
Comment (mwtian) [reply crates/mysten-metrics/src/metered_channel.rs:170]: Dropping the permit and receiving the item will deduct 1 from inflight each. I believe a few Narwhal channel inflight metrics can go negative, which could be caused by this issued.

File: crates/mysten-metrics/src/monitored_mpsc.rs (hunk 1 above)
Comment (mwtian) [reply crates/mysten-metrics/src/monitored_mpsc.rs]: Yes but unfortunately we need to upgrade Rust to use the inspect API.

**Pull request merged**